### PR TITLE
display logs for multiple containers at the same time

### PR DESCRIPTION
**Pull request by baude**
add the ability for users to specify more than one container at a time
while using podman logs.  If more than one container is being displayed,
podman will also prepend a shortened container id of the container on
the log line.

also, enabled the podman-remote logs command during the refactoring of
the above ability.

fixes issue #2219

Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

**Comment by baude** [comment]
@TomasTomecek ptal

File: cmd/podman/logs.go [hunk 1]
```diff
@@ -1,16 +1,13 @@
 package main
 
 import (
-	"os"
+	"github.com/containers/libpod/libpod"
```
Comment (rhatdan) [inline cmd/podman/logs.go]: Bad location.

File: libpod/container_log.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+		cid := l.CID
+		if len(cid) > 12 {
+			cid = cid[:12]
+		}
+		out = fmt.Sprintf("%s ", cid)
+	}
+	if options.Timestamps {
+		out = out + fmt.Sprintf("%s ", l.Time.Format(time.RFC3339Nano))
+	}
+	//if details {
```
Comment (rhatdan) [inline libpod/container_log.go]: What is this?

File: libpod/container_log.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+// Since returns a bool as to whether a log line occurred after a given time
+func (l *LogLine) Since(since time.Time) bool {
+	return l.Time.After(since)
+}
+
+// newLogLine creates a logLine struct from a container log string
+func newLogLine(line string) (*LogLine, error) {
+	splitLine := strings.Split(line, " ")
+	if len(splitLine) < 4 {
+		fmt.Println(line)
```
Comment (rhatdan) [inline libpod/container_log.go]: Debugging?

**Comment by mheon** [comment]
To preserve conversation from IRC, this isn't completely parsing the Kube log format, so it can't interpret P vs F lines and prints extra newlines. Also prints a few timestamp for each character when --follow and --timestamps are specified together.

Handling of control characters also seems wonky.

**Comment by TomasTomecek** [comment]
@baude oh my, thank you so much for working on this! Can't comment on the code, but this seems to be exactly what I was asking for. One question: will `--follow` arg still work?

File: cmd/podman/logs.go [hunk 4]
```diff
@@ -29,6 +26,15 @@ var (
 			logsCommand.GlobalFlags = MainGlobalOpts
 			return logsCmd(&logsCommand)
 		},
+		Args: func(cmd *cobra.Command, args []string) error {
```
Comment (edsantiago) [inline cmd/podman/logs.go]: Nice! Since this now takes multiple args, could you change `Use` to `CONTAINER [CONTAINER...]` ? Thanks.

**Comment by mheon** [comment]
Another thing I noticed here - when following a single container, `podman logs` will exit immediately when the container exits currently, while this will hang forever waiting for new logs.

**Comment by baude** [comment]
> @baude oh my, thank you so much for working on this! Can't comment on the code, but this seems to be exactly what I was asking for. One question: will `--follow` arg still work?

yup!

**Comment by rhatdan** [comment]
I like the idea that it hangs forever, that way if I start the container, the output will start up again.

File: cmd/podman/logs.go [hunk 5]
```diff
@@ -21,17 +18,27 @@ var (
...
 			logsCommand.InputArgs = args
 			logsCommand.GlobalFlags = MainGlobalOpts
 			return logsCmd(&logsCommand)
 		},
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) > 0 && logsCommand.Latest {
+				return errors.New("no containers can be specified when using 'latest'")
+			}
+			if !logsCommand.Latest && len(args) < 1 {
+				return errors.New("specifify at least one container name or ID to log")
```
Comment (TomSweeneyRedHat) [inline cmd/podman/logs.go]: s/specifify/specify

File: libpod/container_log.go [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+	}
+	options.WaitGroup.Add(1)
+	if len(tailLog) > 0 {
+		for _, nll := range tailLog {
+			//nll, err := newLogLine(i)
+			//if err != nil {
+			//	logrus.Error(err)
+			//	continue
+			//}
+			nll.CID = c.ID()
```
Comment (TomSweeneyRedHat) [inline libpod/container_log.go]: can the commented out code be yanked?

**Comment by mheon** [comment]
@baude We can probably remove `pkg/logs` as part of this - that code shouldn't be used elsewhere

**Comment by baude** [comment]
bot, retest this please

**Comment by mheon** [comment]
One more thing I noticed: in the old implementation, `logs --follow` first printed all the logs already there, then began following; with this PR, we don't print anything old, just new log lines as they come up

**Comment by mheon** [comment]
Otherwise, my major concerns are addressed, output looks fine

**Comment by rhatdan** [comment]
bot, retest this please

**Comment by TomasTomecek** [comment]
Nicely done, @baude. Just tested it locally, very nice work!
```
$ ./bin/podman run --rm -d --name cont1 fedora:29 bash -c 'while true; do date; sleep 1; done'
40e06045c6971c4c8452f2780e588ac468942ab3db7fcb9647435d15c8fdc453

$ ./bin/podman run --rm -d --name cont2 fedora:29 bash -c 'while true; do date; sleep 1; done'
2d66875579d62dedb0f0fdfedbeb5ba28dfad7fbe9f2b99b745e17d5f4ce940e

$ ./bin/podman logs -f cont1 cont2
40e06045c697 Fri Mar 15 16:13:43 UTC 2019
40e06045c697 Fri Mar 15 16:13:44 UTC 2019
40e06045c697 Fri Mar 15 16:13:45 UTC 2019
40e06045c697 Fri Mar 15 16:13:46 UTC 2019
40e06045c697 Fri Mar 15 16:13:47 UTC 2019
2d66875579d6 Fri Mar 15 16:13:47 UTC 2019
2d66875579d6 Fri Mar 15 16:13:48 UTC 2019
```

**Comment by baude** [comment]
bot, retest this please

**Comment by mheon** [comment]
Latest version seems to properly print all logs before following

**Pull request merged**